test/Algebra/Graph/Test/NonEmpty/AdjacencyIntMap.hs

```diff
@@ -0,0 +1,626 @@
+{-# LANGUAGE OverloadedLists, ViewPatterns #-}
+-----------------------------------------------------------------------------
+-- |
+-- Module     : Algebra.Graph.Test.NonEmpty.AdjacencyIntMap
+-- Copyright  : (c) Andrey Mokhov 2016-2019
+-- License    : MIT (see the file LICENSE)
+-- Maintainer : andrey.mokhov@gmail.com
+-- Stability  : experimental
+--
+-- Testsuite for "Algebra.Graph.NonEmpty.AdjacencyIntMap".
+-----------------------------------------------------------------------------
+module Algebra.Graph.Test.NonEmpty.AdjacencyIntMap (
+    -- * Testsuite
+    testNonEmptyAdjacencyIntMap
+    ) where
+
+import Control.Monad
+import Data.Semigroup ((<>))
+import Data.Tree
+import Data.Tuple
+
+import Algebra.Graph.NonEmpty.AdjacencyIntMap
+import Algebra.Graph.Test hiding (axioms, theorems)
+import Algebra.Graph.ToGraph (reachable)
+
+import qualified Algebra.Graph.AdjacencyIntMap          as AIM
+import qualified Algebra.Graph.NonEmpty.AdjacencyIntMap as NonEmpty
+import qualified Data.List.NonEmpty                     as NonEmpty
+import qualified Data.Set                               as Set
+import qualified Data.IntSet                            as IntSet
+
+sizeLimit :: Testable prop => prop -> Property
+sizeLimit = mapSize (min 10)
+
+type G = NonEmpty.AdjacencyIntMap
+
+axioms :: G -> G -> G -> Property
+axioms x y z = conjoin
+    [       x + y == y + x                      // "Overlay commutativity"
+    , x + (y + z) == (x + y) + z                // "Overlay associativity"
+    , x * (y * z) == (x * y) * z                // "Connect associativity"
+    , x * (y + z) == x * y + x * z              // "Left distributivity"
+    , (x + y) * z == x * z + y * z              // "Right distributivity"
+    ,   x * y * z == x * y + x * z + y * z      // "Decomposition" ]
+
+theorems :: G -> G -> Property
+theorems x y = conjoin
+    [         x + x == x                        // "Overlay idempotence"
+    , x + y + x * y == x * y                    // "Absorption"
+    ,         x * x == x * x * x                // "Connect saturation"
+    ,             x <= x + y                    // "Overlay order"
+    ,         x + y <= x * y                    // "Overlay-connect order" ]
+
+testNonEmptyAdjacencyIntMap :: IO ()
+testNonEmptyAdjacencyIntMap = do
+    putStrLn "\n============ NonEmpty.AdjacencyIntMap ============"
+    test "Axioms of non-empty graphs"   axioms
+    test "Theorems of non-empty graphs" theorems
+
+    putStrLn $ "\n============ Ord NonEmpty.AdjacencyIntMap ============"
```

Comment:
It would be nice to reduce code duplication for non-empty tests. We could move some tests into `Algebra.Graph.Test.Generic`, adding suffix `NonEmpty` to properties for disambiguatation.

Reply:
`testAdjacencyMap` would be a model to go off? It seems like most of the functions from the `NAM` modules are candidates

Reply:
Yes, I think we can reuse tests for most functions from `NonEmpty` modules. Just thought that perhaps creating `Generic.NonEmpty` would be best since then we don't need to do any renaming.

Reply:
Sorry, I'm not clear about the some of the issues and steps. What would need to be renamed? Does the plan look something like this?
* `Algebra.Graph.Test.API` module should get `NonEmpty` functions added and the nonempty types should get a `nonemptyAPI :: API (G/Mono) (Ord/(~) Int)`
* also add the `ToGraph` instance for NAIM
* new `TestsuiteInt g -> IO ()` functions go in `Algebra.Graph.Test.Generic.NonEmpty`
* each nonempty module would import the `API`, `Generic`, and `Generic.NonEmpty` modules similar to what `Algebra.Graph.Test.AdjacencyMap` does

Reply:
After refreshing my memory on how things are implemented, here is a more concrete plan: 

* Extend `data API g c` with new fields that correspond to missing methods (most of them end with the suffix `1`). In fact there is a TODO for this: https://github.com/snowleopard/alga/blob/026ba0e14f267e0f3cd19feec1e64b2d5e4ecd35/test/Algebra/Graph/Test/API.hs#L58

* Add tests for missing methods to `Test.Generic` or `Test.Generic.NonEmpty` -- whichever turns out to be clearer/more convenient.

* I don't think we need any `ToGraph` instances for non-empty graphs but perhaps I'm mistaken? Instead, I expect `API` definitions similar to https://github.com/snowleopard/alga/blob/026ba0e14f267e0f3cd19feec1e64b2d5e4ecd35/test/Algebra/Graph/Test/API.hs#L146-L148

* Tests for non-empty modules would need to import `API`, `Generic` and perhaps also `Generic.NonEmpty` (if it turns out to be convenient).

Reply:
Thanks for the plan! Refactoring the nonempty modules to use generic is a bit daunting, but I'll try to find time and energy soon

Reply:
@jitwit If you don't want to deal with this daunting task feel free to simply leave a TODO describing the plan. I do understand that it's no fun at all and happy to do it myself :)